light-client/verifier/src/verifier.rs

```diff
@@ -0,0 +1,111 @@
+// Copyright (c) 2018-2023 The MobileCoin Foundation
+
+use crate::{Error, TrustedValidatorSet};
+use mc_blockchain_types::{Block, BlockContents, BlockID, BlockIndex, BlockMetadata};
+use mc_transaction_core::tx::TxOut;
+use std::{collections::HashSet, ops::Range};
+
+/// The light client verifier
+///
+/// This object is capable of:
+/// * Verifying a Block given BlockMetadata containing node signatures
+/// * Verifying one or more TxOut's, that appeared in a given Block.
+///
+/// The verifier does not make network connections and its state does not
+/// change when it verifies things. It needs to be configured with correct
+/// trusted validator sets to give give correct results.
+#[derive(Clone, Debug)]
+pub struct LightClientVerifier {
+    /// A quorum configuration and expected signing keys for the validator
+    /// network.
+    pub trusted_validator_set: TrustedValidatorSet,
+    /// The first block index at which this trusted validator set is applied.
+    pub trusted_validator_set_start_block: BlockIndex,
+    /// A list of historical validator sets, and ranges of block indices at
+    /// which they should be used.
+    pub historical_validator_sets: Vec<(Range<BlockIndex>, TrustedValidatorSet)>,
+    /// A list of known valid block ids, which may appear before
+    /// `trusted_validator_set_start_block` and outside of any of the historical
+    /// ranges.
+    pub known_valid_block_ids: HashSet<BlockID>,
+}
+
+// For a simple LightClientVerifier initialization
+impl From<TrustedValidatorSet> for LightClientVerifier {
+    fn from(src: TrustedValidatorSet) -> Self {
+        Self {
+            trusted_validator_set: src,
+            trusted_validator_set_start_block: 0,
+            historical_validator_sets: Default::default(),
+            known_valid_block_ids: Default::default(),
+        }
+    }
+}
+
+impl LightClientVerifier {
+    /// Verify that a given block has been externalized by the network, given
+    /// additional evidence in the form of BlockMetadata with signatures,
+    /// relative to our configuration of trusted validator sets.
+    pub fn verify_block(&self, block: &Block, metadata: &[BlockMetadata]) -> Result<(), Error> {
+        if !block.is_block_id_valid() {
+            return Err(Error::InvalidBlockId);
+        }
+        if self.known_valid_block_ids.contains(&block.id) {
+            return Ok(());
+        }
+        if block.index >= self.trusted_validator_set_start_block {
+            return self
+                .trusted_validator_set
+                .verify_block_id_signatures(&block.id, metadata);
+        }
+
+        for (range, validator_set) in self.historical_validator_sets.iter() {
+            if range.contains(&block.index) {
+                return validator_set.verify_block_id_signatures(&block.id, metadata);
```

Comment:
Might the ranges overlap? If so, we should try all matching ranges, and return Ok if any of them succeed

Reply:
I don't think it makes sense for the ranges to overlap, that would mean that there are two different authoritative validator sets for some block.

I think it might be okay to test all remaining ranges and return an error if there was overlap, but since I don't think this is going to happen, I don't think it's super important. That could also happen during input validation instead.